src/pudl/transform/eia923.py

```diff
@@ -27,44 +27,27 @@ def _yearly_to_monthly_records(df, md):
     Args:
         df (pandas.DataFrame): A pandas DataFrame containing the annual data to be
             converted into monthly records.
-        md (dict): a dictionary with the integers 1-12 as keys, and the patterns used
-            to match field names for each of the months as values. These patterns are
-            also used to rename the columns in the dataframe which is returned, so they
-            need to match the entire portion of the column name that is month specific.
 
     Returns:
         pandas.DataFrame: A dataframe containing the same data as was passed in via df,
-        but with monthly records instead of annual records.
+        but with monthly records as rows instead of as columns.
 
     """
-    yearly = df.copy()
-    all_years = pd.DataFrame()
-
-    for y in yearly.report_year.unique():
-        this_year = yearly[yearly.report_year == y].copy()
-        monthly = pd.DataFrame()
-        for m in md:
-            # Grab just the columns for the month we're working on.
-            this_month = this_year.filter(regex=md[m]).copy()
-            # Drop this month's data from the yearly data frame.
-            this_year.drop(this_month.columns, axis=1, inplace=True)
-            # Rename this month's columns to get rid of the month reference.
-            this_month.columns = this_month.columns.str.replace(
-                md[m], '', regex=True)
-            # Add a numerical month column corresponding to this month.
-            this_month['report_month'] = m
-            # Add this month's data to the monthly DataFrame we're building.
-            monthly = pd.concat([monthly, this_month], sort=True)
-
-        # Merge the monthly data we've built up with the remaining fields in
-        # the data frame we started with -- all of which should be independent
-        # of the month, and apply across all 12 of the monthly records created
-        # from each of the # initial annual records.
-        this_year = this_year.merge(monthly, left_index=True, right_index=True)
-        # Add this new year's worth of data to the big dataframe we'll return
-        all_years = pd.concat([all_years, this_year], sort=True)
-
-    return all_years
+    ends_with_month_filter = df.columns.str.extract(
+        r'.+_(\w+?)$').isin(set(pc.month_name_to_num_dict.keys())).squeeze()
```

Comment:
Here is a comparison with pandas built-ins.
```
month_names = pd.to_datetime(range(1,13), format='%m').strftime('%B').str.lower()
mask = df.columns.isin(month_names)
...
df.loc[:, 'month_col'] = pd.to_datetime(df.loc[:, 'month_col'], format='%B').dt.month
```
vs
```
month_dict = {'january': 1, 'february': 2, ...}
mask = df.columns.isin(month_dict.keys())
...
df.loc[:, 'month_col'] = df.loc[:, 'month_col'].map(month_dict)
```
I don't have a strong preference, but I think the dictionary version is more readable. I do think the month dict should be directly in the function though, rather than hidden in pudl.constants (there aren't any other references to it).